cli/src/template_builder.rs

```diff
@@ -1116,6 +1116,18 @@ fn builtin_functions<'a, L: TemplateLanguage<'a> + ?Sized>() -> TemplateBuildFun
             content, labels,
         ))))
     });
+    map.insert(
+        "plain_text",
+        |language, diagnostics, build_ctx, function| {
+            let [content_node] = function.expect_exact_arguments()?;
+            let content =
+                expect_template_expression(language, diagnostics, build_ctx, content_node)?;
+            let template = ReformatTemplate::new(content, move |formatter, recorded| {
+                write!(formatter, "{}", String::from_utf8_lossy(recorded.data()))
+            });
+            Ok(L::wrap_template(Box::new(template)))
+        },
```

Comment:
Perhaps, we can instead add `stringify()` function of `Template -> String` type? It will be a thin wrapper of `expect_plain_text_expression()`.

Reply:
I just tried using expect_plain_text_expression directly, I wonder if this is simpler?

Reply:
Yes, that's what I meant. No `Template` wrapper is needed. It can also propagate error.

Let's also rename the function since it's no longer a `Template -> Template` function.

Reply:
Ah, I see what you mean. Sorry, I thought you were talking about adding a `stringify` method on the Rust Template type, so was confused.

Re: name, not a strong opinion but `plain_text` still reads better to me than `stringify` -- especially in indicating that it strips formatting (by ignoring labels). `stringify` could also be misleading when used on expressions evaluated to strings (for example, `commit.change_id().normal_hex()` returns a string but with formatting from the "inherited" change_id label).

Reply:
> I thought you were talking about adding a `stringify` method on the Rust Template type, so was confused.

Ah, that could be an option, too, but I don't recommend it because `non_template_type.stringify()` doesn't work.

> Re: name

Hmm, I assumed another use case is `stringify(<template>).<string-method>(..)`, and `plain_text()` wouldn't make much sense here. I don't have a strong feeling about that, though.

FWIW, `commit.change_id().normal_hex()` is technically equivalent to `label("commit change_id normal_hex", stringify(commit.change_id().normal_hex()))`.

Reply:
Okay, I think these are 2 different use cases -- one to convert content to plain text and the other to rich text, so we could either have plain_text and rich_text functions (not necessarily with those names) or a single stringify function that takes an optional, default false plain_text param. WDYT?

Reply:
They're the same under the hood. There's no "rich text" thing in template because a template String can't include color codes.

Reply:
> [...] template String can't include color codes.

I'm confused -- this is exactly why I needed this PR in the first place, right? `change_id.normal_hex()` already returns a string but one that includes color codes?

Reply:
It's a tricky bit, but method calls have implicit `label()`s. `change_id.normal_hex()` is desugared to `label("change_id normal_hex", change_id.normal_hex())`. The `String` property itself doesn't have colors. If it had, color codes would be sanitized (as you mentioned in the other thread.)

BTW, if we add `raw(content: String-ish)` function to bypass sanitization, it will effectively drop color labels within `content`. So you can probably write `raw("\e]8;;" ++ url ++ "\e\\")` without using `plain_text(url)`.

Reply:
> It's a tricky bit, but method calls have implicit label()s.

Yes, I only realized this yesterday when I was going through the code to debug why there's color codes in `change_id.normal_hex()`.

> The String property itself doesn't have colors. If it had, color codes would be sanitized (as you mentioned in the other thread.)

Yes, what I'd meant was there could be a template that evaluates to a string (i.e., string methods like `len` work on it) _and_ have color codes (like `change_id.normal_hex()`).

As for `rich_text`, I don't have a use case for it, but consider someone wants to trim `content` (string or template) to just the first line or display it in uppercase etc., _while retaining the existing labels_ -- AFAICT, this is not possible today. But if there was a `rich_text` function, they could do `rich_text(content).first_line()` or `rich_text(content).upper()` etc. -- I'm not sure how hard it would be to implement this though (to get methods like `substr` right).

---

> BTW, if we add raw(content: String-ish) function to bypass sanitization, it will effectively drop color labels within content. So you can probably write raw("\e]8;;" ++ url ++ "\e\\") without using plain_text(url).

Yes, I was also thinking about this -- what do you think about this function being called `plain_text` instead of `raw`? `raw` could be confused with the already supported raw string literals and `plain_text` also aligns with the behavior of the existing PlainTextFormatter (in that, it doesn't sanitize any escape codes).

Either way, I think we have some alignment here to just add one new function that takes a string-ish template (might as well convert any template to string, I guess?) and writes it unsanitized, correct?

Reply:
> Yes, what I'd meant was there could be a template that evaluates to a string (i.e., string methods like `len` work on it) _and_ have color codes (like `change_id.normal_hex()`).

Perhaps, `ShortestIdPrefix` belongs to this category. It's string-like, but has its own rendering logic. For general-purpose text transformation, we'll probably need to add `<template>.first_line()` or `first_line(template)`. The latter would be somewhat similar to `indent()`.

> > BTW, if we add raw(content: String-ish) function to bypass sanitization, it will effectively drop color labels within content. So you can probably write raw("\e]8;;" ++ url ++ "\e\") without using plain_text(url).
> 
> Yes, I was also thinking about this -- what do you think about this function being called `plain_text` instead of `raw`? `raw` could be confused with the already supported raw string literals and `plain_text` also aligns with the behavior of the existing PlainTextFormatter (in that, it doesn't sanitize any escape codes).

`plain_text()` doesn't sound like a function to bypass sanitizing (and potentially mess up output coloring.) I don't know if `raw()` is better, but it at least implies a "use at your own risk" thingy.

> Either way, I think we have some alignment here to just add one new function that takes a string-ish template (might as well convert any template to string, I guess?) and writes it unsanitized, correct?

Yes, I think that's the easiest way to print escape code transparently.

Reply:
I _think_ I more or less addressed this thread now -- I renamed the function to raw_text. PTAL! I'll add docs once the implementation looks good to you. Thanks!